source/base/parameter_handler.cc

```diff
@@ -2311,13 +2319,391 @@ ParameterHandler::print_parameters (std::ostream     &out,
     }
 
   // dive recursively into the subsections
-  print_parameters_section (out, style, 0);
+  recursively_print_parameters (std::vector<std::string>(), // start at the top level
+                                style,
+                                0,
+                                out);
 
   return out;
 }
 
 
 
+void
+ParameterHandler::recursively_print_parameters (const std::vector<std::string> &target_subsection_path,
+                                                const OutputStyle               style,
+                                                const unsigned int              indent_level,
+                                                std::ostream                   &out) const
+{
+  AssertThrow (out, ExcIO());
+
+  Assert ((style != XML) && (style != JSON),
+          ExcMessage ("This function should not be necessary for XML or JSON output."));
+
+  const boost::property_tree::ptree &current_section
+    = entries->get_child (collate_path_string(target_subsection_path));
+
+  unsigned int overall_indent_level = indent_level;
+
+  switch (style)
+    {
+    case Text:
+    case ShortText:
+    {
+      // first find out the longest entry name to be able to align the
+      // equal signs to do this loop over all nodes of the current
+      // tree, select the parameter nodes (and discard sub-tree nodes)
+      // and take the maximum of their lengths
+      std::size_t longest_name = 0;
+      for (boost::property_tree::ptree::const_iterator
+           p = current_section.begin();
+           p != current_section.end(); ++p)
```

Comment:
For me the following is working:

```
for (const auto p: current_section)
  if (is_parameter_node (p.second) == true)
    longest_name = std::max (longest_name, 
                             demangle(p.first).length());
```
Actually, you just have to switch `p->` to `p.` in your second try.